### ISSUE TITLE
Infinite recursion in net-stubbing/lib/server/middleware/request.js 

### ISSUE THREAD

**Issue by eirikur-grid** [cypress-io/cypress#22693]
### Current behavior

We run our cypress tests on CircleCI, distributing the tests among several runners. Sometimes we get a failure like this one:
```
RangeError: Maximum call stack size exceeded
    at RegExp.get source [as source] (<anonymous>:null:null)
    at RegExp.toString (<anonymous>:null:null)
    at equalByTag (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:5786:35)
    at baseIsEqualDeep (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:3357:13)
    at baseIsEqual (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:3316:14)
    at equalObjects (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:5872:43)
    at baseIsEqualDeep (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:3375:14)
    at baseIsEqual (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:3316:14)
    at baseIsMatch (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:3433:19)
    at /root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:3603:37
    at baseFindIndex (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:816:11)
    at Function.findIndex (/root/.cache/Cypress/9.6.0/Cypress/resources/app/node_modules/lodash/lodash.js:7322:14)
    at getRouteForRequest (/root/.cache/Cypress/9.6.0/Cypress/resources/app/packages/net-stubbing/lib/server/route-matching.js:105:77)
    at populateMatchingRoutes (/root/.cache/Cypress/9.6.0/Cypress/resources/app/packages/net-stubbing/lib/server/middleware/request.js:45:67)
    at populateMatchingRoutes (/root/.cache/Cypress/9.6.0/Cypress/resources/app/packages/net-stubbing/lib/server/middleware/request.js:50:13)
    at populateMatchingRoutes (/root/.cache/Cypress/9.6.0/Cypress/resources/app/packages/net-stubbing/lib/server/middleware/request.js:50:13)
... and on and on and on
```

### Desired behavior

Test should run without crashing

### Test code to reproduce

I haven't found a reliable way to reproduce this. It's not always the same test that fails. However, the failures I've seen so far occur in tests with a `beforeEach` function that calls a utility method, which uses `cy.intercept` for a lot of our API routes.

### Cypress Version

9.6.0

**Comment by mjhenkes** [comment]
Closing absent a reliable reproduction.

Please open a new issue with a reproducible example and link to this issue. Here are some tips for providing a [Short, Self Contained, Correct, Example](http://sscce.org/) and our own [Troubleshooting Cypress](https://on.cypress.io/troubleshooting) guide.

**Comment by eirikur-grid** [comment]
I've been looking at the source code and I suspect I've found the root cause. 
In route-matching.ts there's a function named `getRouteForRequest`. It accepts an optional `prevRoute` object, from which to start its search. This is achieved by:
```javascript
const possibleRoutes = prevRoute ? orderedRoutes.slice(_.findIndex(orderedRoutes, prevRoute) + 1) : orderedRoutes
```
Looking at the [lodash documentation](https://lodash.com/docs/#findIndex) for the `findIndex` function, I've surmised that it performs a deep comparison of objects, matching by value rather than reference. Hence, if there are two identical routes, it will match the first one, even if `prevRoute` is a reference to the second one.

I don't have my work laptop with me so I'm unable to whip up a repro right now.
